src/Processors/Formats/IRowInputFormat.cpp

```diff
@@ -122,6 +122,26 @@ Chunk IRowInputFormat::generate()
             }
         }
     }
+    catch (ParsingException & e) 
+    {
+        String verbose_diagnostic;
+        try
+        {
+            verbose_diagnostic = getDiagnosticInfo();
+        }
+        catch (const Exception & exception)
+        {
+            verbose_diagnostic = "Cannot get verbose diagnostic: " + exception.message();
+        }
+        catch (...)
+        {
+            /// Error while trying to obtain verbose diagnostic. Ok to ignore.
+        }
+
+        e.setLineNumber(total_rows);
+        e.addMessage("(at row {})\n" + verbose_diagnostic);
```

Comment:
Maybe it would be easier to make the ParsingException add a fixed-form detail message to the end of the exception message, like when using `addExtendedMessage`. It would look like "Some complicated parser error: row 1234". This way you don't have to remember to put an extra `{}` into the message.

Reply:
Sure.

Reply:
Maybe it can add this exta `{}` in constructor then? Now the caller has to remember to call a special `addMessage`, and only if the line number is not zero -- this looks error-prone.

Reply:
I think there should always be a `{}` after main message, and we should print there either `(at row 1234)` or empty string.